### PR TITLE
[Admin][Zones] Choosing disabled country as a zone member made possible for admin user

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.11           |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| License         | MIT                                                          |

Right now we’re experiencing a bug-o-feature, where when you disable a country that is a zone member, the members list on the Zone shows empty dropdown and field is unable to be controlled.

Changes made:

- adding disabled countries as zone members made possible for admin users
- search dropdown added instead of select dropdown field

Before:
502e0587-5f2a-46e6-bf0a-978a51e0462c
After:
Zrzut ekranu 2022-06-7 o 15 24 15

### CODE REVIEW

File: src/Sylius/Bundle/AddressingBundle/Tests/Form/Type/CountryChoiceTypeTest.php [hunk 1]
```diff
@@ -87,32 +92,56 @@ public function it_returns_all_countries(): void
...
+        $this->countryRepository->findBy(['enabled' => true])->willReturn([
+            $this->france->reveal(),
+            $this->poland->reveal(),
+        ]);
+
+        $this
+            ->assertChoicesLabels(['Poland'], ['choice_filter' => static fn (?CountryInterface $country): bool => $country !== null && $country->getName() === 'Poland',
+            'enabled' => true,
+            'multiple' => true
+        ]);
```
Comment (coldic3) [inline src/Sylius/Bundle/AddressingBundle/Tests/Form/Type/CountryChoiceTypeTest.php:142]:
```suggestion
        $this
            ->assertChoicesLabels(['Poland'], [
                'choice_filter' => static fn (?CountryInterface $country): bool => $country !== null && $country->getName() === 'Poland',
                'enabled' => true,
                'multiple' => true
            ])
        ;
```

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 2]
```diff
@@ -41,14 +41,24 @@ public function configureOptions(OptionsResolver $resolver): void
...
-                    if (null === $options['enabled']) {
-                        $countries = $this->countryRepository->findAll();
-                    } else {
+                    if (null !== $options['enabled'] && $options['multiple']) {
                         $countries = $this->countryRepository->findBy(['enabled' => $options['enabled']]);
+                    } else {
+                        $countries = $this->countryRepository->findAll();
                     }
 
                     return $countries;
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php:50]:
```suggestion
                    if (null !== $options['enabled'] && $options['multiple']) {
                        $return $this->countryRepository->findBy(['enabled' => $options['enabled']]);
                    }

                    return $this->countryRepository->findAll();
```

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 3]
```diff
@@ -41,14 +41,24 @@ public function configureOptions(OptionsResolver $resolver): void
             ->setDefaults([
                 'choice_filter' => null,
                 'choices' => function (Options $options): iterable {
-                    if (null === $options['enabled']) {
-                        $countries = $this->countryRepository->findAll();
-                    } else {
+                    if (null !== $options['enabled'] && $options['multiple']) {
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php:44]: Instead of changing the logic here by adding a condition with `multiple` here, you can probably set the proper value of `enabled` in the appropriate form type

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 4]
```diff
@@ -41,14 +41,24 @@ public function configureOptions(OptionsResolver $resolver): void
...
                 },
+                'attr' => static function (Options $options): array {
+                    if (!$options['multiple']) {
+
+                        return ['class' => 'sylius_zone_member_country_select ui fluid search selection dropdown'];
+                    }
+                    else {
+
+                        return [];
+                    }
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php:60]: 👍 however I have mixed feelings about it generally. We're providing strictly UI- and SemanticUI-oriented data (names of classes) in the form type which theoretically could be used in any other template :dancer: We need to think about how to safely move from this nonsense (without BC breaks). Still, nothing to change now 🖖

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 5]
```diff
@@ -41,18 +41,16 @@ public function configureOptions(OptionsResolver $resolver): void
             ->setDefaults([
                 'choice_filter' => null,
                 'choices' => function (Options $options): iterable {
-                    if (null === $options['enabled']) {
-                        $countries = $this->countryRepository->findAll();
-                    } else {
-                        $countries = $this->countryRepository->findBy(['enabled' => $options['enabled']]);
+                    if ($options['enabled'] === true) {
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php:44]:
```suggestion
                    if ($options['enabled']) {
```

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 6]
```diff
@@ -41,18 +41,16 @@ public function configureOptions(OptionsResolver $resolver): void
...
                     }
 
-                    return $countries;
+                    return $this->countryRepository->findAll();
                 },
                 'choice_value' => 'code',
                 'choice_label' => 'name',
                 'choice_translation_domain' => false,
-                'enabled' => true,
+                'enabled' => false,
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php:53]: Revert this change and leave it as it was before not to change the default behaviour, because this `CountryChoiceType` can be used in many places. Instead of changing it here, set the proper value of this field in the appropriate form: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php#L104

File: src/Sylius/Bundle/AddressingBundle/Tests/Form/Type/CountryChoiceTypeTest.php [hunk 7]
```diff
@@ -49,6 +52,12 @@ protected function setUp(): void
         $poland->getName()->willReturn('Poland');
         $this->poland = $poland;
 
+        /** @var ProphecyInterface|CountryInterface $austria */
+        $austria = $this->prophesize(CountryInterface::class);
+        $austria->getCode()->willReturn('AT');
+        $austria->getName()->willReturn('Austria');
+        $this->austria = $austria;
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Tests/Form/Type/CountryChoiceTypeTest.php:59]: Shouldn't this country be disabled for the tests?

File: src/Sylius/Bundle/CoreBundle/Form/Extension/ChannelTypeExtension.php [hunk 8]
```diff
@@ -57,6 +57,7 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                 'label' => 'sylius.form.channel.countries',
                 'required' => false,
                 'multiple' => true,
+                'enabled' => true,
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Form/Extension/ChannelTypeExtension.php:60]: Shouldn't be needed after my above suggestions

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php (hunk 5 above)
Comment (TheMilek) [reply src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php]: Actually I would stay with checking equality, because without it, it is also checking existance of the key, so for example
`'enabled' => false` will be in this case true

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php [hunk 9]
```diff
@@ -30,6 +31,16 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
         $builder->addModelTransformer(new ReversedTransformer(new ResourceToIdentifierTransformer($this->countryRepository, 'code')));
     }
 
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver
+            ->setDefaults([
+                'attr' => ['class' => 'country_search_dropdown ui fluid search selection dropdown'],
+                'enabled' => 'false',
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php:39]: I can add here almost the same comment as in the previous iteration: https://github.com/Sylius/Sylius/pull/14058#discussion_r891936298

`CountryCodeChoiceType` is used in several different places if I'm correct, but you want to change the behaviour only for selecting countries for zones, so this parameter should be disabled only in `ZoneType`, at least for now

Comment (GSadee) [reply src/Sylius/Bundle/AddressingBundle/Form/Type/CountryCodeChoiceType.php]: I'm almost sure that if you will change this line: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/AddressingBundle/Form/Type/ZoneType.php#L104 by adding there an `enabled` parameter, it will work correctly:
```
ZoneInterface::TYPE_COUNTRY => ['label' => 'sylius.form.zone.types.country', 'enabled' => false],
```
And you will not need to add a custom form type. Did you check this approach?

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Kamil! :tada:
